src/Analyzers/CSharp/Analyzers/UseIsNullCheck/CSharpUseIsNullCheckOverIsObject.cs

```diff
@@ -0,0 +1,83 @@
+    {
+        public CSharpUseIsNullCheckOverIsObject()
+            : base(IDEDiagnosticIds.UseIsNullOverIsObjectDiagnosticId,
+                   EnforceOnBuildValues.UseIsNullCheck,
+                   CSharpCodeStyleOptions.PreferIsNullCheckOverIsObject,
+                   CSharpAnalyzersResources.Use_is_null_check,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Null_check_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                var objectType = context.Compilation.ObjectType;
+                if (objectType.TypeKind == TypeKind.Error)
+                {
+                    return;
+                }
+
+                // All trees should have the same language version. Bail-out early in compilation start instead of checking every tree.
+                var tree = context.Compilation.SyntaxTrees.FirstOrDefault();
+                if (tree is null || ((CSharpParseOptions)tree.Options).LanguageVersion < LanguageVersion.CSharp9)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(c => AnalyzeOperation(c, objectType), OperationKind.IsType, OperationKind.NegatedPattern);
+            });
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol objectType)
+        {
+            var option = context.Options.GetOption(CSharpCodeStyleOptions.PreferIsNullCheckOverIsObject, context.Operation.Syntax.SyntaxTree, context.CancellationToken);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            if (ShouldReportDiagnostic(context.Operation, objectType))
+            {
+                var severity = option.Notification.Severity;
+                context.ReportDiagnostic(
+                    DiagnosticHelper.Create(
+                        Descriptor, context.Operation.Syntax.GetLocation(), severity, additionalLocations: null, properties: null));
+            }
+        }
+
+        private static bool ShouldReportDiagnostic(IOperation operation, INamedTypeSymbol objectType)
+        {
+            if (operation is IIsTypeOperation isTypeOperation)
+            {
+                return objectType.Equals(isTypeOperation.TypeOperand, SymbolEqualityComparer.Default);
+            }
+            else if (operation is INegatedPatternOperation negatedPattern)
+            {
+                return negatedPattern.Pattern is ITypePatternOperation typePatternOperation &&
+                    objectType.Equals(typePatternOperation.MatchedType, SymbolEqualityComparer.Default);
+            }
```

Comment:
perhaps show examples here of each.

Reply:
@CyrusNajmabadi Not sure what you mean here. Can you please clarify? Thanks!

Reply:
i mean, put a comment like `// this matches: x is y z` to make it clearer what cases we're matching. up to you though.  i personally liek seeing it though it can be argued it is redundant with the code.